server/src-lib/Hasura/Class.hs

```diff
@@ -0,0 +1,84 @@
+module Hasura.Class
+  ( MetadataStorageT(..)
+  , runMetadataStorageT
+  , MonadMetadataStorage(..)
+  )
+where
+
+import           Hasura.Eventing.HTTP
+import           Hasura.Eventing.ScheduledTrigger.Types
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+
+import           Control.Monad.Morph                    (MFunctor, hoist)
+
+import qualified Hasura.Tracing                         as Tracing
+
+newtype MetadataStorageT m a
+  = MetadataStorageT {unMetadataStorageT :: ExceptT QErr m a}
+  deriving ( Functor, Applicative, Monad
+           , MonadError QErr
+           , MFunctor
+           , MonadTrans
+           , MonadIO
+           )
+
+runMetadataStorageT
+  :: MetadataStorageT m a -> m (Either QErr a)
+runMetadataStorageT =
+  runExceptT . unMetadataStorageT
+
+
+class (Monad m) => MonadMetadataStorage m where
+
+  -- Scheduled triggers
+  getDeprivedCronTriggerStats :: MetadataStorageT m [CronTriggerStats]
+  getScheduledEventsForDelivery :: MetadataStorageT m ([CronEvent], [OneOffScheduledEvent])
+  insertScheduledEvent :: ScheduledEventSeed -> MetadataStorageT m ()
+  insertScheduledEventInvocation
+    :: Invocation 'ScheduledType -> ScheduledEventType -> MetadataStorageT m ()
+  setScheduledEventOp
+    :: ScheduledEventId -> ScheduledEventOp -> ScheduledEventType -> MetadataStorageT m ()
+  unlockScheduledEvents
+    :: ScheduledEventType -> [ScheduledEventId] -> MetadataStorageT m Int
+  unlockAllLockedScheduledEvents :: MetadataStorageT m ()
```

Comment:
@0x777 and myself discussed a potential future design here that avoids spelling out all of these methods explicitly. TODO: expand this sentence.

Reply:
Ideally we would've liked to avoid having functions that are specific to scheduled/cron triggers and instead have functions that provide a generic 'event storage and retrieval' interface but we'll have to change a lot of existing code for scheduled and cron triggers. We can get to this after the multi-source work is done. (@rakeshkky maybe you can put this as a comment under TODO in `MonadMetadataStorage` class's description.